Index: Changelog.txt
===================================================================
RCS file: /ctbot/ct-Bot/Changelog.txt,v
retrieving revision 1.84
diff -u -r1.84 Changelog.txt
--- Changelog.txt	14 Jun 2006 09:57:21 -0000	1.84
+++ Changelog.txt	21 Jun 2006 17:10:49 -0000
@@ -1,5 +1,10 @@
 CHANGELOG fuer c't-Bot
 ======================
+2006-06-21	Achim Pankalla (achim.pankalla@gmx.de) 
+			Die Funktion sensor_abstand() nun auch beim ct-bot.exe aktiv. Anpassung an den 
+			neuen ct-Sim der nun nicht mehr Abstandswerte in mm liefert, sondern simulierte
+			GP2D12 Analogwerte.
+
 2006-06-13	Torsten Evers (tevers@onlinehome.de) Anzeige des aktuell freien RAMs auf Screen 5 hinzugefuegt (nur MCU!)
 
 2006-06-12	Benjamin Benz (bbe@heise.de) Promo8-FB-Codes erweitert
Index: command.c
===================================================================
RCS file: /ctbot/ct-Bot/command.c,v
retrieving revision 1.13
diff -u -r1.13 command.c
--- command.c	19 May 2006 14:04:26 -0000	1.13
+++ command.c	21 Jun 2006 17:10:50 -0000
@@ -43,6 +43,8 @@
 #include "bot-logik.h"
 #include "bot-2-pc.h"
 
+#include "sensor-low.h"
+
 #include <stdio.h>
 #include <string.h>
 
@@ -303,8 +305,8 @@
 		// Einige Kommandos ergeben nur fuer simulierte Bots Sinn
 		#ifdef PC
 			case CMD_SENS_IR:
-				sensDistL=received_command.data_l;
-				sensDistR=received_command.data_r;
+				sensDistL=sensor_abstand(received_command.data_l);
+				sensDistR=sensor_abstand(received_command.data_r);
 				#ifdef TIME_AVAILABLE
 					system_time_isr();		/* Einmal pro Update-Zyklus aktualisieren wir die Systemzeit */
 				#endif
Index: include/sensor-low.h
===================================================================
RCS file: /ctbot/ct-Bot/include/sensor-low.h,v
retrieving revision 1.2
diff -u -r1.2 sensor-low.h
--- include/sensor-low.h	3 Feb 2006 10:54:42 -0000	1.2
+++ include/sensor-low.h	21 Jun 2006 17:10:50 -0000
@@ -25,6 +25,13 @@
 #ifndef sens_low_H_
 #define sens_low_H_
 
+/*! 
+ * Funktion die versucht, die Daten der Abstandssensoren auf mm zu linearisieren.
+ * @param sensor_data Die Daten vom analogen Sensor
+ * @return Distanz in mm
+ */
+extern int16 sensor_abstand(int16 sensor_data);
+
 /*!
  * Initialisiere alle Sensoren
  */
Index: include/sensor_correction.h
===================================================================
RCS file: /ctbot/ct-Bot/include/sensor_correction.h,v
retrieving revision 1.2
diff -u -r1.2 sensor_correction.h
--- include/sensor_correction.h	8 Jun 2006 16:36:38 -0000	1.2
+++ include/sensor_correction.h	21 Jun 2006 17:10:50 -0000
@@ -25,7 +25,12 @@
 #ifndef SENSOR_CORRECTION_H_
 #define SENSOR_CORRECTION_H_
 
-#define SENSDISTSLOPE	  65330		/*!< Steigung der Sensorkennlinie */
+#ifdef MCU
+	#define SENSDISTSLOPE	  65330		/*!< Steigung der Sensorkennlinie */
+#endif 
+#ifdef PC
+	#define SENSDISTSLOPE	  42240		/*!< Steigung der Sensorkennlinie */
+#endif 
 #define SENSDISTOFFSET		-3		/*!< Offset der Sensorkennlinie */
 
 /* Parameter fuer die IR-Sensoren*/
Index: pc/sensor-low_pc.c
===================================================================
RCS file: /ctbot/ct-Bot/pc/sensor-low_pc.c,v
retrieving revision 1.2
diff -u -r1.2 sensor-low_pc.c
--- pc/sensor-low_pc.c	3 Feb 2006 10:54:42 -0000	1.2
+++ pc/sensor-low_pc.c	21 Jun 2006 17:10:50 -0000
@@ -27,7 +27,9 @@
 
 #ifdef PC
 
+#include "sensor.h"
 #include "sensor-low.h"
+#include "sensor_correction.h"
 
 /*!
  * Initialisiere alle Sensoren
@@ -35,6 +37,18 @@
 void bot_sens_init(void){
 }
 
+/*! 
+ * Funktion die versucht, die Daten der Abstandssensoren auf mm zu linearisieren.
+ * @param sensor_data Die Daten vom analogen Sensor
+ * @return Distanz in mm
+ */
+int16 sensor_abstand(int16 sensor_data){
+	// TODO reale Kennlinie beachten!!!!
+	return SENSDISTSLOPE / (sensor_data - SENSDISTOFFSET);
+	
+//	return 1023-sensor_data;
+}
+
 /*!
  * Alle Sensoren aktualisieren.
  * Das geschieht auf der PC Seite anders -- in einem eigenen Thread.